[Custom Directive] Add post about GraphQL custom directives (#600)
artsy/artsy.github.io: _posts/2020-01-03-graphql-custom-directives.markdown

to better handle errors occurring during query resolution, and better equip clients to reason about such errors.

The goal is to describe our current approach, but also do a deep dive into specific ways we've extended our
[GraphQL server](artsy/metaphysics) to help us accomplish that. Therefore, if you are an interested GraphQL user,
you may find this useful, even if some of the larger context specifically around how we are using it to help
standardize error handling doesn't apply.

<!-- more -->

## Introduction and Initial Context

At Artsy, we use GraphQL as our API language of choice. In particular, [as we've described before](link to prev
blogposts), we have an orchestration layer speaking GraphQL, which is what our front-end clients talk to. The
GraphQL orchestration layer wraps up access to several backend services, which are made accessible via a
combination of data loaders and [GraphQL stitching](apollo link). [Apollo Federation](apoll link) is another tool
people are using to bring together disaparte backends when using GraphQL in an orchestration layer.

Now, consider the following query, which is a realistic one you might see when accessing a 'product' page.

> ```

Reviewer: these ticks can accept a graphql type to highlight text (including down below)

Author: Interesting! Is that the case? Looks like the list of [supported languages](https://pygments.org/languages/) doesn't include GraphQL, and other posts in the blog don't use tags for GraphQL fragments.

Will update other snippets here tho (typescript, json, etc)

The goal is to describe our current approach, but also do a deep dive into specific ways we've extended our
[GraphQL server](https://github.com/artsy/metaphysics) to help us accomplish that. Therefore, if you are an
interested GraphQL user, you may find this useful, even if some of the larger context specifically around how we
are using it to help standardize error handling doesn't apply.

<!-- more -->

## Introduction and Initial Context

At Artsy, we use GraphQL as our API language of choice. In particular,
[as](https://artsy.github.io/blog/2018/05/08/is-graphql-the-future/)
[we've described](https://artsy.github.io/blog/2016/06/19/graphql-for-mobile/)
[before](https://artsy.github.io/blog/2016/11/02/improving-page-speed-with-graphql/), we have an orchestration
layer speaking GraphQL, which is what our front-end clients talk to. The GraphQL orchestration layer wraps up
access to several backend services, which are made accessible via a combination of data loaders and
[schema stitching](https://www.apollographql.com/docs/apollo-server/features/schema-stitching/).
[Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) is another tool

> people are using to bring together disaparte backends when using GraphQL in an orchestration layer.

Reviewer 2: ```suggestion
people are using to bring together disparate backends when using GraphQL in an orchestration layer.
```

to the user, and the UI should be able to gracefully handle when it's missing. Since we use React/Relay/GraphQL,
and GraphQL queries are colocated with UI components, a GraphQL directive is particularly useful.

Rewriting the above query, we might do something like:

```javascript
{
  artwork(id: "artwork") {
    mainContentStuff @principalField
    biographicalData
    userReviews {
        ...
    }  # Accesses a back-end reviews service
    ...
  }
}
```

This would mean that any errors occuring in resolving `mainContentStuff` would result in either a 500 status code
and error page to the user, or possibly a more specific error and status code. However, any errors occurring in
resolving user reviews or other fields, would not cause a 500 and error page. Instead, there would be a 200 and the
UI would render. This means that our UI components should generally be defensive about their incoming props being
`null` (which is likely what you'd see when the corresponding field errors during query resolution). Using
TypeScript and strict null checking can help make your UI bulletproof to these sorts of issues.

> So, now that some of the context is here, let's look at how we implement a custom `@principalField` GraphQL

Reviewer 3: ```suggestion
With this context, let's look at how we implement a custom `@principalField` GraphQL
```

directive. Future posts in this series talking about custom GraphQL functionality will likely skip this intro
section.

## Implementing a Custom GraphQL Directive

First, we have a new directive that we'd like to add to our schema. That is, we'd like a client to be able to
specify `@principalField` alongside any field, and have that query be validated correctly by the server. You can
[see for yourself](<https://metaphysics-staging.artsy.net/?query=%7B%0Aartwork(id%3A%22andy-warhol-skull%22)%20%40nonExistentDirective%7B%0A%20%20id%0A%7D%7D>)
how [graphql-js](https://github.com/graphql/graphql-js) and
[express-graphql](https://github.com/graphql/express-graphql) respond when an unknown directive is specified. Your
GraphQL server implementation may look slightly different.

Turns out, that's pretty easy to do in `graphql-js`. We declare a variable of type `GraphQLDirective`, which
operates on a `DirectiveLocation.FIELD` location:

```javascript
const PrincipalFieldDirective = new GraphQLDirective({
  name: "principalField",
  locations: [DirectiveLocation.FIELD]
})
```

and then when we create our schema, we pass this in as `directives`. Since this will overwrite the [standard

> directives](link to spec), we need to append ours.

Reviewer 3: #todo

Something like:

```javascript
import { specifiedDirectives } from "graphql"

new GraphQLSchema({
  directives: [...specifiedDirectives, PrincipalFieldDirective],
  query: ...
  ...
})
```

In terms of the SDL for your schema, this is equivalent to the following line:

```
directive @principalField on FIELD
```

That's it! You've successfully added a new directive to your schema. At this point, your GraphQL server will
properly validate and allow a `@principalField` directive specified by a client alongside any field. Now, we do
want to ensure that if this directive is used, it only appears once in your query. We can accomplish this with a

> custom GraphQL validation, which we'll cover in the next post.

Reviewer 3: Reminder to link here when the next post is available.

So now, how should we implement the functionality of this directive? We decided that we want to use the
[extensions](https://github.com/graphql/graphql-spec/blob/master/spec/Section%207%20--%20Response.md#response-format)
part of our GraphQL response to carry this data, a free-form map of data up to the implementor, which is a perfect
fit for this type of optional additional information. If an error occurs in a field tagged with the directive, we
want the response to look something like:

```json
{

>   data: {

Reviewer: If it's json it will need quotes :)

Author: heh, got me!

        path.push(name)
      },
      leave() {
        path.pop()
      }
    },
    Directive(node) {
      if (node.name.value === "principalField") {
        return BREAK
      }
    }
  })

  return path
}
```

This uses a [GraphQL visitor](https://graphql.org/graphql-js/language/#visit) to traverse our query, and build up
an array of field names. We can exit early with that path if we encounter our directive.

## Example Query

Let's take a look at how you can use this in practice, in order to help standardize when and with what status a UI
can inform the user of an error.

Something like:

```
{
  artwork(id: "andy-warhol-skull") @principalField {
    userReviews {
      notes
    }
    contents
    ...
  }
}
```

results in:

```json
{
  data: {
    artwork: null
  }
  extensions: {
    principalFieldError: {
      httpStatusCode: 404
    }
  }
}
```

and the UI can immediately return an appropriate message to the user, if the artwork is not found. If fetching the
artwork is successful, but there's an issue with the reviews, the response will look like:

```json
{
  data: {
    artwork: {
      userReviews: null
      contents: ...
    }
  }
}
```

> We can make sure that our UI component (likely Relay containers) that corresponds to `userReviews` is defensive

Reviewer 3: ```suggestion
We can make sure that our UI components (likely Relay containers) corresponding to `userReviews` are defensive
```

about that incoming prop being `null`. Most likely a zero state ("No Reviews Found"), or just skipping the section
entirely, is appropriate.

## Conclusion

In this way, we can standardize on and remove ambiguity about how a UI handles one or more errors in query
resolution and exactly when such an error should be propagated and made user-facing. We can help ensure that our
UI's are resilient to errors occurring in a leaf.

In the next post, we'll look at how we can write a GraphQL validation rule to ensure that a client specifies at
most one field with the `@principalField` directive. After that, we'll take a look at how you can successfully
parse and support GraphQL query resolution errors occurring in a variety of contexts such as during stitching.

> Putting this altogether, we hope you will come away with a better understanding for how to extend your GraphQL

Reviewer 3: ```suggestion
Putting this altogether, we hope you will come away with a better understanding of how to extend your GraphQL
```
